### PR TITLE
blip-0039: BOLT 11 Invoice Blinded Path Tagged Field

### PR DESCRIPTION
**Pull request by ellemouton**
This bLIP defines a new tagged field, `b`,  for the payment invoice
encoding described by BOLT 11 which can be used to communicate 
encoded blinded path information to the payer of the invoice. It also 
introduces a new feature bit for the `I` context to indicate the 
presence of the new field in an invoice. 

Delving Post: https://delvingbitcoin.org/t/blip-bolt-11-invoice-blinded-path-tagged-field/991
Implementation: https://github.com/lightningnetwork/lnd/pull/8752

### CODE REVIEW

File: blip-0039.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+provide. Once the Offers protocol along with its new invoice format has been 
+widely deployed, then there will be no use for this BOLT 11 carve-out. Due to 
+the forcasted temporary use of the new field, it makes sense to be in bLIP form
+rather than adding this in a more temporary way to the spec via a BOLT update 
+proposal.
+
+## Backwards Compatibility
+
+BOLT 11 states that the reader of an invoice "MUST skip over unknown fields". 
+This means that an un-updated reader of an invoice that includes the new tagged
```
[the remark this answers is not in the document]
Comment (t-bast) [reply blip-0039.md:134]: Agreed, if you use a blinded path, but it's optional and the invoice contains identifying information for the recipient to fallback when not using blinded paths, then that blinded path doesn't provide any privacy?

[the remark this answers is not in the document]
Comment (t-bast) [reply blip-0039.md:134]: Or maybe I got this wrong: there is no routing hint anyway, so a reader that doesn't know how to read the blinded path won't have a complete enough invoice and will simply fail without attempting the payment?

File: blip-0039.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+and using them in the context of receiving payments is natively supported in 
+the [BOLT 12 Offers proposal][offers]. However, the Offers proposal also 
+includes various other dependencies such as onion messaging and various new 
+protocol message all of which will require a network wide upgrade before full 
+advantage can be taken of the new protocol. Blinded paths themselves are a 
+useful tool for privacy and potentially more reliable payment delivery due to 
+receiver being able to select paths it knows to be more reliable. This document 
+proposes a carve-out to the existing [BOLT 11][bolt11] invoice format so that 
+advantage can be taken of the blinded paths protocol in the context of payments 
+in a sooner time frame. This will be done by adding a new tagged-field type to
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply blip-0039.md:33]: > Sure, just that everyone but lnd already did that so the total time there is now zero :)

There's still quite a lot of existing work unfinished w.r.t the long tail of ensuring blinded paths works well in production, eg: 
   * how many introduction nodes 
   * how are they chosen, path finding algo changes, etc
   * dummy hop concerns
   * mpp over blinded paths
   * CLTV and other policy buffers
   * fixed path lengths

Having a set invoice format allows implementations to hone in on the questions above in isolation. Eg: IIUC some implementations only support creating/sending to a single blinded path. For us being able to test blinded paths in isolation has enabled us to find some interesting existing edge cases in spec/implementation, and also find some new higher order use cases. 

> For all intents and purposes its a new invoice format

That's quite an interpretation of the phrase "invoice format" 🙃

File: blip-0039.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply blip-0039.md:134]: > Or maybe I got this wrong: there is no routing hint anyway, so a reader that doesn't know how to read the blinded path won't have a complete enough invoice and will simply fail without attempting the payment?

Yeah there would be no routing hint, a required feature bit, and also the pubkey in the invoice would be random (not actually the node ID of the node creating the invoice).

File: blip-0039.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ellemouton) [reply blip-0039.md]: updated this to say that it is for testing 👍

File: blip-0039.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+  invoice containing the `b` field since this value is expected to be accounted 
+  for in each path's accumulated `cltv_expiry_delta`.
+- The existing invoice `expiry` field along with the `timestamp` field  
+  should be used to communicate the `max_cltv_expiry` of the blinded paths in 
+  the invoice. The reader should use the 10-minutes-per-block assumption to 
+  calculate an estimation of the `max_cltv_exipiry` value.
+
+## Universality
+
+This proposal is a temporary measure that will allow users to start making use
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply blip-0039.md:122]: added a piece at the end re this being for testing - lemme know if that looks ok?

File: blip-0039.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ellemouton) [reply blip-0039.md:135]: added some extra info at the end - lemme know if this looks ok

**Review by ellemouton — COMMENTED** [review]
Addressed the concerns :)

File: blip-0002.md [hunk 4]
```diff
@@ -45,11 +45,12 @@ Custom feature bits used in the `I` [Bolt 11](https://github.com/lightning/bolts
...
-| 258/259     | `dns_resolver`         | This node accepts DNSSEC proof requests           | N                |                                  | [bLIP 32](./blip-0032.md)        |
+| Bits    | Name                  | Description                                              | Context | Dependencies            | Link                      |
+|---------|-----------------------|----------------------------------------------------------|---------|-------------------------|---------------------------|
+| 54/55   | `keysend`             | A form of spontaneous payment                            | N       | `var_onion_optin`       | [bLIP 3](./blip-0003.md)  |
+| 256/257 | `hosted_channels`     | This node accepts requests for hosted channels           | IN      |                         | [bLIP 17](./blip-0017.md) |
+| 258/259 | `dns_resolver`        | This node accepts DNSSEC proof requests                  | N       |                         | [bLIP 32](./blip-0032.md) |
+| 260/261 | `bolt11_blinded_path` | This invoice may contain a new blinded path tagged field | I       | `option_route_blinding` | [bLIP 39](./blip-0039.md) |
```
Comment (carlaKC) [inline blip-0002.md:53]: Conflicts with the feature bits in #27!

Comment (ellemouton) [reply blip-0002.md:53]: oooo great catch! thank you!

**Comment by ellemouton** [comment]
updated to 262/263 - thanks @carlaKC 🙏

**Review by Roasbeef — APPROVED** [review]
LGTM 🎍
